vignettes/type-promotion.Rmd

```diff
@@ -0,0 +1,106 @@
+---
+title: "Type Promotion"
+output: rmarkdown::html_vignette
+vignette: >
+  %\VignetteIndexEntry{Type Promotion}
+  %\VignetteEngine{knitr::rmarkdown}
+  %\VignetteEncoding{UTF-8}
+---
+
+```{r, echo = FALSE}
+knitr::opts_chunk$set(
+  eval = !(Sys.getenv("PJRT_PLATFORM", "cpu")  == "metal")
+)
+```
+
+## Type Promotion Rules
+
+When combining tensors of different types (e.g., adding an `f32` to an `i32`), {anvil} needs to determine a common type.
+For example, below we are adding an `f32` to an `f64`, where the former is promoted to the latter's type, because it's more expressive.
+
+```{r}
+library(anvil)
+jit(nv_add)(
+  nv_scalar(1.0, dtype = "f32"),
+  nv_scalar(1.0, dtype = "f64")
+)
+```
+
+The type-promotion rules are inspired by JAX, and they are designed for execution on accelerators like GPUs, where one often wants speed instead of precision.
+
+The rules are defined by the `common_dtype()` function.
+
+```{r}
+common_dtype(dt_f64, dt_f32)
```

Comment:
nit: the `dt_` prefix here reminds me more of dates than data types. I'd rather just use strings, eg `common_dtype("f64", "f32")` if possible.